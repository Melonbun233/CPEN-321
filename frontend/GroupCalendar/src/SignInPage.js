'use strict';
//This file is for the login page
//We will also implement sign up here

import React, {Component} from 'react';
import {Text, View, StyleSheet, KeyboardAvoidingView,
		Alert, Button, ScrollView} 
		from 'react-native';
import { GoogleSignin, statusCodes} from 'react-native-google-signin';
import {TextField} from 'react-native-material-textfield';
import Ripple from 'react-native-material-ripple';
import Network from './common/GCNetwork';
import cs from './common/CommonStyles';
import Storage from './common/Storage';
import * as config from './../config.json';

export default class SignInPage extends Component {
	//this page dont need a header
	static navigationOptions = {
		header: null
	}

	constructor(props) {
		super(props);
	
		this.state = {
			//checking whether user has signed in
			isChecking: true,
			//user is signing by app
			isLoading: false,
			//user is signing by google
			isSigning: false, 
			userEmail: '',
			userPwd: '',
			errors: {},
		};

		//general callback when focusing text field
		this._onFocus = this._onFocus.bind(this);

		this._onSubmitEmail = this._onSubmitEmail.bind(this);
		this._onSubmitPassword = this._onSubmitPassword.bind(this);

		//setting up references
		this.emailRef = this._updateRef.bind(this, 'email');
		this.passwordRef = this._updateRef.bind(this, 'password');
		this._onSignInButtonPressed = this._onSignInButtonPressed.bind(this);
		this._onGoogleSignInPressed = this._onGoogleSignInPressed.bind(this);
	}

	//only called once
	//we will check whether user has signed out
	//if not, we will just sign in automatically
	async componentDidMount() {
		GoogleSignin.configure(config.googleSignIn);
		await this.checkUserSignedIn();
	}

	//check whether user has signed in
	async checkUserSignedIn(){
		try {
			let profile = await Storage.getProfile();
			if (profile == null){
				//user hasn't signed in
				this.setState({isChecking: false});
			} else {
				this.setState({isChecking: false});
				this.props.navigation.navigate('Main');
			}
		} catch (error) {
			Alert.alert('Something Went Wrong');
			//user hasn't signed in
			this.setState({isChecking: false});
		}
	}

	_onFocus(){
		let {errors} = this.state;
		for (let key in errors) {
			let ref = this[key];
			if(ref.isFocused()){
				delete errors[key];
			}
		}
		this.setState({errors});
	}

	_updateRef(name, ref){
		this[name] = ref;
	}

	_onSubmitEmail(){
		this.password.focus();
	}

	_onSubmitPassword(){
		this.password.blur();
		this._onSignInButtonPressed();
	}

	//Function handles sign in button press
	// we will send email and password to 
	async _onSignInButtonPressed(){
		let {userEmail, userPwd} = this.state;
		//we first get the user info by the username
		this.setState({isLoading: true});
		try {
			let status = await Network.verifyUser(userEmail, userPwd);
			//Alert.alert(JSON.stringify(status));
			switch (status) {
				//correct user_email and user_pwd
				//we save the user info to async storage and jump to main pages
				case 200: {
					this.setState({isLoading: false});
					Storage.setSignInByGoogle('false');
					//jump to main page
					this.props.navigation.navigate('Main');
				}
				break;
				case 400:
				case 404: this.setState(
						{errors: {
							email: 'incorrect email or password',
							password: 'incorrect email or password',
						}});
				break;
				default: Alert.alert('Internet Error', JSON.stringify(res.error));
			}
		} catch (error) {
			Alert.alert('Something went wrong');
		}
		this.setState({isLoading: false});
	}

	//sign in by google
	//on successful signin, we send the google id token to server and get 
	//our cookie
	async _onGoogleSignInPressed(){
		this.setState({isSigning: true});
		await GoogleSignin.hasPlayServices();
		await GoogleSignin.signIn()
			.then(async (userInfo) => 
			{
				let status = await Network.verifyUserByGoogle(userInfo);
				switch(status) {
					//already has an account
					case 200: {
						Storage.setSignInByGoogle('true');
						this.setState({isSigning: false});
						this.props.navigation.navigate('Main');
					}
					break;
					//doesnt has a password
					case 0: {
						this.props.navigation.push('GoogleSignIn');
					}
					break;
					default: {
						Alert.alert('Internet Error', JSON.stringify(status));
					}
				}
			})
			.catch((error) => {
				if (error.code === statusCodes.SIGN_IN_CANCELLED) {
					//user canceled the login flow
				} else if (error.code === statusCodes.IN_PROGRESS) {
					//in progress
				} else if (error.code === statusCodes.PLAY_SERVICES_NOT_AVAILABLE) {
					// play services not available or outdated
					Alert.alert('Play Service Not Avaliable');
				} else {
					//some other error happened
					Alert.alert('Something went wrong\n', JSON.stringify(error));
				}
			});
		this.setState({isSigning: false});
	}

	render() {
		let {isSigning, isLoading, isChecking, errors} = this.state;
		if (isChecking) {
			return (
				<View></View>
			);
		}
		//let {errors} = this.state;
		return (
			<KeyboardAvoidingView 
				behavior="padding" 
				style = {[cs.container, cs.wholePage, s.scrollContainer]}
			>
			<ScrollView
				style = {[s.scrollContainer]}
				keyboardShouldPersistTaps = 'never'
				scrollEnabled = {false}
			>
				{/*sign up button*/}
				<View style = {[cs.container, s.signUpContainer]}>
					<Button 
						testID = 'SignUpButton'
						title = 'Sign up'
						color = '#66a3ff'
						onPress = {() => this.props.navigation.push('SignUp')}
					/>
				</View>
				{/*title*/}
				<View style = {[cs.container, s.titleContainer]}>
					<Text style = {cs.title}>Group</Text>
					<Text style = {cs.title}>Calendar</Text>
					<Text style = {[cs.h3, s.welcome]}>Welcome Back</Text>
				</View>
				{/*user email and password*/}
				<View style = {[s.contentContainer]}>
					<TextField
						testID = 'UserEmail'
						ref = {this.emailRef}
						label = 'Email'
						fontSize = {18}
						labelHeight = {24}
						onChangeText = {(text) => this.setState({userEmail: text})}
						autoCorrect = {false}
						autoCapitalize = 'none'
						onSubmitEditing = {this._onSubmitEmail}
						returnKeyType = 'next'
						keyboardType = 'email-address'
						textContentType = 'username'
						error = {errors.email}
						onFocus = {this._onFocus}
					/>
					<TextField
						testID = 'UserPwd'
						ref = {this.passwordRef}
						fontSize = {18}
						labelHeight = {24}
						label = 'Password'
						inputContainerPadding = {4}
						onChangeText = {(text) => this.setState({userPwd: text})}
						secureTextEntry= {true}
						keyboardType = 'default'
						returnKeyType = 'go'
						onSubmitEditing = {this._onSubmitPassword}
						textContentType = 'password'
						clearTextOnFocus = {true}
						error = {errors.password}
						onFocus = {this._onFocus}
					/>
				</View>

			{/*sign in buttons*/}
				
				<Ripple
					testID = 'AppSignInButton'
					disabled = {isLoading}
					onPress = {this._onSignInButtonPressed}
					style = {[cs.container, s.buttonContainer]}
				>
						<Text style = {s.buttonMsg}>
							{isLoading ? 'Signing in...' : 'Sign in'}
						</Text>
				</Ripple>

				<Ripple
					testID = 'GoogleSignInButton'
					onPress = {this._onGoogleSignInPressed}
					disabled = {isSigning} 
					style = {[cs.container, s.buttonContainer]}
				>
						<Text style = {s.buttonMsg}>Sign in by Google</Text>
				</Ripple>
				
			</ScrollView>
			</KeyboardAvoidingView>
			);
	}
} 



const s = StyleSheet.create({
	welcome: {
		color: '#e6e6e6',
	},
	scrollContainer: {
		paddingTop: 10,
		flex: 1,
		width: '100%',
		height: '100%',
	},
	signUpContainer: {
		flex: 1,
		width: '100%',
		paddingRight: 20,
		paddingTop: 30,
		alignItems: 'flex-end',
	},
	titleContainer: {
		marginLeft: '10%',
		justifyContent: 'flex-start',
		alignItems: 'flex-start',
		flex: 1,
	},
	contentContainer: {
		marginLeft: '10%',
		width: '80%',
		flex: 1,
	},
	buttonContainer: {
		flex: 1,
		marginLeft: '10%',
		marginTop: 5,
		marginBottom: 5,
		width: '80%',
		height: 40,
		backgroundColor: '#66a3ff',
	},
	buttonMsg: {
		color : '#ffffff',
		fontSize: 18,
	},
	failLogin: {
		alignItems: 'center',
		fontSize: 14,
		color: '#ff0000',
	},
});