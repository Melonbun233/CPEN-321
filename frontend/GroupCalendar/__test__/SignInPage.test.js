import React from 'react';
import SignInPage from '../src/SignInPage';
import {shallow, render} from 'enzyme';
import sinon from 'sinon';

//mock navigation and spy on it
const navigation = {
	navigate : jest.fn(),
	push : sinon.spy(),
}

describe('SignInPage test', () => {
	const checkUserSignedInSpy = sinon.spy(SignInPage.prototype, 'checkUserSignedIn');
	const appSignInSpy = sinon.spy(SignInPage.prototype, '_onSignInButtonPressed');
	const googleSignInSpy = sinon.spy(SignInPage.prototype, '_onGoogleSignInPressed');
	const focusSpy = sinon.spy(SignInPage.prototype, '_onFocus');
	const submitEmailSpy = sinon.spy(SignInPage.prototype, '_onSubmitEmail');
	const submitPasswordSpy = sinon.spy(SignInPage.prototype, '_onSubmitPassword');
	var wrapper;

	beforeEach(() => {
		wrapper = shallow(<SignInPage 
			navigation = {navigation}
		/>, { lifecycleExperimental: true });

		wrapper.setState({isChecking: false,});
		wrapper.instance().password = {
			blur: jest.fn(),
			focus: jest.fn(),
		};
	});

	test('renders SignInPage', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('component did mount', () => {
		wrapper.instance().checkUserSignedIn();
		expect(checkUserSignedInSpy.callCount).toBe(1);
	});

	test('update ref', () => {
		wrapper.instance()._updateRef('test', 'testRef');
		expect(wrapper.instance().test).toBe('testRef');
	});

	test('change user email', () => {
		wrapper.find({testID: 'UserEmail'}).simulate('ChangeText', 'testEmail');
		expect(wrapper.instance().state.userEmail).toBe('testEmail');
	});

	test('change user password', () => {
		wrapper.find({testID: 'UserPwd'}).simulate('ChangeText', 'testPwd');
		expect(wrapper.instance().state.userPwd).toBe('testPwd');
	});
	
	test('focus text field', () => {
		wrapper.find('TextField')
			.forEach((child) => {
				child.simulate('Focus');
			})
		expect(focusSpy.callCount).toBe(2);
	})

	test('submit editing email', () => {
		wrapper.find({testID: 'UserEmail'}).simulate('SubmitEditing');
		expect(submitEmailSpy.callCount).toBe(1);
	});

	test('submit editing password', () => {
		wrapper.find({testID: 'UserPwd'}).simulate('SubmitEditing');
		expect(submitPasswordSpy.callCount).toBe(1);
		expect(appSignInSpy.callCount).toBe(1);
	});

	test('click App sign in buttons', () => {
		wrapper.find({testID: 'AppSignInButton'}).simulate("Press");
		expect(appSignInSpy.callCount).toBe(2);
	});

	test('click Google sign in buttons', () => {
		wrapper.find({testID: 'GoogleSignInButton'}).simulate("Press");
		expect(googleSignInSpy.callCount).toBe(1);
	});

	test('click sign up buttons', () => {
		wrapper.find({testID: 'SignUpButton'}).simulate("Press");
		expect(navigation.push.callCount).toBe(1);
	});
});