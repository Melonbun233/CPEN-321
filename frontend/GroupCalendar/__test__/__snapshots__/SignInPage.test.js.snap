// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInPage test renders SignInPage 1`] = `
<KeyboardAvoidingView
  behavior="padding"
  enabled={true}
  keyboardVerticalOffset={0}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
      },
      Object {
        "height": "100%",
        "width": "100%",
      },
      Object {
        "flex": 1,
        "height": "100%",
        "paddingTop": 10,
        "width": "100%",
      },
    ]
  }
>
  <ScrollView
    keyboardShouldPersistTaps="never"
    scrollEnabled={false}
    style={
      Array [
        Object {
          "flex": 1,
          "height": "100%",
          "paddingTop": 10,
          "width": "100%",
        },
      ]
    }
  >
    <Component
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          },
          Object {
            "alignItems": "flex-end",
            "flex": 1,
            "paddingRight": 20,
            "paddingTop": 30,
            "width": "100%",
          },
        ]
      }
    >
      <Button
        color="#66a3ff"
        onPress={[Function]}
        testID="SignUpButton"
        title="Sign up"
      />
    </Component>
    <Component
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          },
          Object {
            "alignItems": "flex-start",
            "flex": 1,
            "justifyContent": "flex-start",
            "marginLeft": "10%",
          },
        ]
      }
    >
      <Component
        style={
          Object {
            "color": "#000000",
            "fontSize": 60,
            "fontWeight": "bold",
          }
        }
      >
        Group
      </Component>
      <Component
        style={
          Object {
            "color": "#000000",
            "fontSize": 60,
            "fontWeight": "bold",
          }
        }
      >
        Calendar
      </Component>
      <Component
        style={
          Array [
            Object {
              "color": "#000000",
              "fontSize": 30,
              "fontWeight": "bold",
            },
            Object {
              "color": "#e6e6e6",
            },
          ]
        }
      >
        Welcome Back
      </Component>
    </Component>
    <Component
      style={
        Array [
          Object {
            "flex": 1,
            "marginLeft": "10%",
            "width": "80%",
          },
        ]
      }
    >
      <TextField
        activeLineWidth={2}
        animationDuration={225}
        autoCapitalize="none"
        autoCorrect={false}
        baseColor="rgba(0, 0, 0, .38)"
        disableFullscreenUI={true}
        disabled={false}
        disabledLineType="dotted"
        disabledLineWidth={1}
        editable={true}
        errorColor="rgb(213, 0, 0)"
        fontSize={18}
        inputContainerPadding={8}
        keyboardType="email-address"
        label="Email"
        labelFontSize={12}
        labelHeight={24}
        labelPadding={4}
        lineWidth={0.5}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        returnKeyType="next"
        testID="UserEmail"
        textColor="rgba(0, 0, 0, .87)"
        textContentType="username"
        tintColor="rgb(0, 145, 234)"
        titleFontSize={12}
        underlineColorAndroid="transparent"
      />
      <TextField
        activeLineWidth={2}
        animationDuration={225}
        autoCapitalize="sentences"
        baseColor="rgba(0, 0, 0, .38)"
        clearTextOnFocus={true}
        disableFullscreenUI={true}
        disabled={false}
        disabledLineType="dotted"
        disabledLineWidth={1}
        editable={true}
        errorColor="rgb(213, 0, 0)"
        fontSize={18}
        inputContainerPadding={4}
        keyboardType="default"
        label="Password"
        labelFontSize={12}
        labelHeight={24}
        labelPadding={4}
        lineWidth={0.5}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        returnKeyType="go"
        secureTextEntry={true}
        testID="UserPwd"
        textColor="rgba(0, 0, 0, .87)"
        textContentType="password"
        tintColor="rgb(0, 145, 234)"
        titleFontSize={12}
        underlineColorAndroid="transparent"
      />
    </Component>
    <Ripple
      disabled={false}
      onPress={[Function]}
      onRippleAnimation={[Function]}
      rippleCentered={false}
      rippleColor="rgb(0, 0, 0)"
      rippleContainerBorderRadius={0}
      rippleDuration={400}
      rippleFades={true}
      rippleOpacity={0.3}
      rippleSequential={false}
      rippleSize={0}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          },
          Object {
            "backgroundColor": "#66a3ff",
            "flex": 1,
            "height": 40,
            "marginBottom": 5,
            "marginLeft": "10%",
            "marginTop": 5,
            "width": "80%",
          },
        ]
      }
      testID="AppSignInButton"
    >
      <Component
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 18,
          }
        }
      >
        Sign in
      </Component>
    </Ripple>
    <Ripple
      disabled={false}
      onPress={[Function]}
      onRippleAnimation={[Function]}
      rippleCentered={false}
      rippleColor="rgb(0, 0, 0)"
      rippleContainerBorderRadius={0}
      rippleDuration={400}
      rippleFades={true}
      rippleOpacity={0.3}
      rippleSequential={false}
      rippleSize={0}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          },
          Object {
            "backgroundColor": "#66a3ff",
            "flex": 1,
            "height": 40,
            "marginBottom": 5,
            "marginLeft": "10%",
            "marginTop": 5,
            "width": "80%",
          },
        ]
      }
      testID="GoogleSignInButton"
    >
      <Component
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 18,
          }
        }
      >
        Sign in by Google
      </Component>
    </Ripple>
  </ScrollView>
</KeyboardAvoidingView>
`;
